### PR TITLE
Optimize union case fields enumeration

### PR DESCRIPTION
**Pull request by auduchinok**
`UnionCase.RecdFields` needlessly converts arrays to lists in few cases where it's only used to get fields count during type check. This PR replaces these few places with comparing initial arrays lengths.

### CODE REVIEW

File: src/fsharp/symbols/Symbols.fs [hunk 1]
```diff
@@ -781,9 +781,13 @@ and FSharpUnionCase(cenv, v: UnionCaseRef) =
         checkIsResolved()
         v.Range
 
+    member __.HasFields =
```
Comment (TIHan) [inline src/fsharp/symbols/Symbols.fs]: This seems like it would be useful, but it's never used yet. I would make another PR with `HasFields` and a use-case for it.

**Comment by TIHan** [comment]
It seems like the intent was for `RecdFields` to be immutable, but it's more efficient to use an array but it's not immutable. I wish we could just use `ImmutableArray` instead as it's literally what we want. I really really do not like seeing two APIs, `RecdFields`, `RecdFieldsArray`.

@dsyme perhaps we should start using `ImmutableArray` and if Fable has some issues, we can help them as well?

**Comment by dsyme** [comment]
@tihan I'm OK in principle with using an immutable array type in the compiler. We should, I think, do it systematically in a separate PR? 

There is an open, approved language suggestion to add an immutable array type to FSharp.Core, we could also do it via that route. See the suggestion for discussion about naming and overlap with S.C.Immutable.

File: src/fsharp/symbols/Symbols.fs (hunk 1 above)
Comment (auduchinok) [reply src/fsharp/symbols/Symbols.fs]: There are many FCS APIs that are used outside of this repo only. What kind of use case you want me to add?

**Comment by auduchinok** [comment]
Making it and similar things immutable arrays would be great, though it's probably going to make some places less readable due to not using list pattern matching the way it's done now.

**Comment by TIHan** [comment]
@dsyme yes, it should be a separate PR. I'm fine with the current change as is.

File: src/fsharp/symbols/Symbols.fs (hunk 1 above)
Comment (TIHan) [reply src/fsharp/symbols/Symbols.fs]: Gotcha. Perhaps a test would be fine for it.

Comment (auduchinok) [reply src/fsharp/symbols/Symbols.fs]: OK, I've extracted this change to another PR and I'll a test there.
